net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +65,668 @@ impl UdpSocket {
         Self::bind_raw(addr)
     }
 
+    /// Is the socket broken and needs a rebind?
+    pub fn is_broken(&self) -> bool {
+        self.is_broken.load(std::sync::atomic::Ordering::SeqCst)
+    }
+
+    /// Marks this socket as needing a rebind
+    pub fn mark_broken(&self) {
+        self.is_broken
+            .store(true, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    /// Rebind the underlying socket.
+    pub fn rebind(&self) -> Result<()> {
+        debug!("rebinding {}", self.addr);
+        // Remove old socket
+        let mut guard = self.socket.write().unwrap();
+        {
+            let Some(socket) = guard.take() else {
+                bail!("cannot rebind closed socket");
+            };
+            drop(socket);
+        }
+
+        // Prepare new socket
+        let new_socket = inner_bind(self.addr)?;
+
+        // Insert new socket
+        guard.replace(new_socket);
+
+        // Clear errors
+        self.is_broken
+            .store(false, std::sync::atomic::Ordering::SeqCst);
+
+        Ok(())
+    }
+
     fn bind_raw(addr: impl Into<SocketAddr>) -> Result<Self> {
-        let addr = addr.into();
-        let network = IpFamily::from(addr.ip());
-        let socket = socket2::Socket::new(
-            network.into(),
-            socket2::Type::DGRAM,
-            Some(socket2::Protocol::UDP),
-        )
-        .context("socket create")?;
-
-        if let Err(err) = socket.set_recv_buffer_size(SOCKET_BUFFER_SIZE) {
-            warn!(
-                "failed to set recv_buffer_size to {}: {:?}",
-                SOCKET_BUFFER_SIZE, err
-            );
+        let mut addr = addr.into();
+        let socket = inner_bind(addr)?;
+        // update to use selected port
+        addr.set_port(socket.local_addr()?.port());
+
+        Ok(UdpSocket {
+            socket: RwLock::new(Some(socket)),
+            addr,
+            is_broken: AtomicBool::new(false),
+        })
+    }
+
+    /// Use the socket
+    pub fn with_socket<F, T>(&self, f: F) -> std::io::Result<T>
+    where
+        F: FnOnce(&tokio::net::UdpSocket) -> T,
+    {
+        let guard = self.socket.read().unwrap();
+        let Some(socket) = guard.as_ref() else {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::BrokenPipe,
```

Comment:
how was this error chosen?  i mean, why is this one the right error?  might be worth documenting in the code.

Reply:
it was the closest to me that indicates that the connection is broken, would love to hear better suggestions.